crypto/bls_core.c

```diff
@@ -525,9 +525,10 @@ void bls_batchVerify(const int sigs_len, byte* results, const ep2_st* pks_input,
         if ( read_ret != RLC_OK || check_membership_G1(&sigs[i]) != VALID) {
             if (read_ret == UNDEFINED) // unexpected error case 
                 goto out;
-            // set signature as infinity and set result as invald
+            // set signature and key to infinity (no effect on the aggregation tree)
+            // and set result to invalid
             ep_set_infty(&sigs[i]);
-            ep2_copy(&pks[i], (ep2_st*) &pks_input[i]);
+            ep2_set_infty(&pks[i]);
```

Comment:
🤔 the fix looks great, a few questions out of curiosity: 
The original code (i.e., red one) means that even if the `i-th` signature is found to be invalid during the conversion process, the `i-th` public key in the `pks` array would still hold the original public key from `pks_input` array. 
Instead, the change this PR introduces means that if a signature is found to be invalid, its public key in the `pks` array is set to the point at infinity (i.e., the identity element in ECC). In this way, public keys with an invalid signature do not affect the end result of verification. Now the question is: does it mean that the previous state of the code could affect the subsequent aggregation and verification processes? as keeping the original public key for invalid signatures could lead to incorrect results or vulnerabilities in the batch verification process.

Reply:
Good question!
With the previous code, I concluded there isn't a vulnerability (output `true` for a certain `signature[i]` while it should be false) but found it not straightforward to explain why. The reason is that the public key `key[i]` still interferes with the computation in the greedy first step (greedy because it attempts to aggregate all signatures and all keys at once).
With the new code, `key[i]` is replaced by identity and it doesn't interfere anymore with the greedy step and I saw it's easier to explain why the result is sound. 
In both codes (old and new), the randomization of the keys and signatures using random scalars chosen by the verifier (and not known to the signer) is a key element in why this works. 

I also forgot to mention another performance advantage of the new code (I'll add it to the description above):
 - in the old code, if a signature has a bad format, the greedy step fails, and we have to recurse over individual signatures to see which one is valid and which one is not. 
 - in the new code, if a signature has a bad format, the greedy step may succeed and we could return early.